<a href="https://colab.research.google.com/github/HarshKapr/Cloud/blob/main/SIG742%20Mid%20Term%20Assignment%20Harsh%20Kapoor.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

**StudentId -  224893883**

**Student Name - Harsh Kapoor**

**SIG742 Mid Term Assignment**

**Part - I**

**Answer 1.1**  is_divisible method  takes two integers, m and n. The method returns True if m is divisible by n, and returns False otherwise.

In [ ]:
# write a method is_divisble which takes two integers m and n as input
def is_divisible(m, n):
    return m % n == 0

# Test cases for is_divisible
print(is_divisible(8, 2))  # True as 8 is divisible by 2 and remainder is 0
print(is_divisible(11, 3))  # False as 11 is divisible by 3 and remainder is 2
print(is_divisible(15, 4))  # False as 15 is divisible by 4 and remainder is 3
print(is_divisible(18,6)) # True as 18 is divisible by 6 and remainder is 0



True
False
False
True


**Algo used to solve 1.1**

**Input:** Two integers, m and n.

**Output:** True if m is divisible by n, False otherwise.

**Algorithm:** Calculate the remainder of the division of m by n using the modulo operator (%). If the remainder is 0, return True. Otherwise, return False.


**Answer 1.2** - **“not_equal”** that takes two variables(any variable type) and gives the same result as the != operator.

In [ ]:
# write a method not_equal which takes two integers a and b as input
def not_equal(a, b):
    return not (a == b)

# Test cases for not_equal
print(not_equal(3, 3))  # Expected Output : False
print(not_equal(5, 3))  # Expected Output : True
print(not_equal([1, 2, 3], [1, 2, 3]))  # Expected output: False
print(not_equal('hello', 'harsh'))  # Expected Output : True


False
True
False
True


**Algo used to solve 1.2**

**Input:** Two parameters, a and b.

**Output:** True if a is not equal to b, False otherwise.

**Algorithm:**

Using the equality operator (==) to check if a is equal to b.
Apply the not operator to the result. If a is not equal to b, the result is True; otherwise, it is False.


**Answer 1.3**

In [ ]:
# three coins flip with two possible outcome of [H,T]
results = [(x, y, z) for x in ['H', 'T'] for y in ['H', 'T'] for z in ['H', 'T']]
print(results)

[('H', 'H', 'H'), ('H', 'H', 'T'), ('H', 'T', 'H'), ('H', 'T', 'T'), ('T', 'H', 'H'), ('T', 'H', 'T'), ('T', 'T', 'H'), ('T', 'T', 'T')]


**Expected Outcome -** Each coin flip has two possible outcomes: heads (H) or tails (T). For three coin flips, we need to consider all combinations of these outcomes. There are 2 * 2 * 2 =8 possible outcome as results.

In [ ]:
[x + y for x in [10, 20, 30] for y in [1, 2, 3]] # run the expression as given in question, this will add the elements between two list

[11, 12, 13, 21, 22, 23, 31, 32, 33]

This list comprehension iterates over each element in the first list [10, 20,30] and adds it to each element in the second list [1, 2, 3]. The result is a new list containing all possible sums of these pairs.

The list of comprehension iterators will add the each element of second list.

In [ ]:
# For nested loop which will give same result as quation -- [x+y for x in [10,20,30] for y in [1,2,3]]

result = []
for x in [10, 20, 30]:
    for y in [1, 2, 3]:
        result.append(x + y)
print(result)

[11, 12, 13, 21, 22, 23, 31, 32, 33]


**Answer 1.4** - function find_digits

In [ ]:
def find_digits(s):
    # List comprehension that iterates over each character in the string 's'
    # Checks if the character is a digit using 'isdigit()' method
    # Converts the digit character to an integer using 'int(char)' and adds it to the list
    return [int(char) for char in s if char.isdigit()]

# Test cases to validate the function
print(find_digits("Harsh123"))      # Expected output: [1, 2, 3]
print(find_digits("CRICKET"))   # Expected output: []
print(find_digits("1a2b3c8"))     # Expected output: [1, 2, 3, 8]

[1, 2, 3]
[]
[1, 2, 3, 8]


**Explanation**

Function Definition: 'def find_digits(s):'

This line defines a feature 'find_digits' that takes a unmarried argument 's', that is anticipated to be a string.

List Comprehension: 'return [int(char) for char in s if char.Isdigit()]'

Iteration: 'for char in s' iterates over every man or woman inside the string 's'.

Condition: 'if char.Isdigit()' tests whether the individual char is a digit.

The 'isdigit()' method returns True if the individual is a digit (0-9), in any other case False.

Conversion: 'int(char)' converts the person to an integer if it's miles a digit.

Resulting List: The list comprehension gathers all of the integers into a listing and returns it.

**Test Cases:**

'print(find_digits("Harsh123"))' returns '[1, 2, 3]' because the string consists of the digits 1, 2, and three.

'print(find_digits("Cricket"))' returns '[]' due to the fact there are no digits within the string.

'print(find_digits("1a2b3c8"))' returns '[1, 2, 3, 8]' because these are the digits present inside the string.



**Answer 1.5**

In [ ]:
## List Comprehension to find all [x,y] pairs satisfying the equation within the given domain and range

xy_pairs = [(x,2*x**2-1) for x in range (-3,4) if 0 <=2*x**2-1<=30]

**Explanation:** range (-3,4) generates the integers from -3 to 3 inclusive.
2*x**2 -1 computes the y value for each x.
The if condition filters out any [x,y] pairs where y does not lie with in the range [0,30].

**Set Up the List Comprehension:** Loop through all integer values of (X) within the domain [-3,3]. Calculate the corresponding (y) value using the equation y=2x**2 -1 .
Include the ([x,y]) pair in the list if the computed (y) lies within the range [0,30].

**Conditions:** Ensure that the calculated y values satisfies the condition 0<=y<=30.
Use the range () function to iterate through the integers for x.


**Answer 1.6**

For this first, we have to calculate the Euclidean distance between the centers of the two balls, and check If it is less than or equal to the sum of their radii, if it is less than or equal than the balls will be colliding.

Given both balls collide if distance between centers is less than or equal to the sum of their radii.

Here is the Euclidean distance formula to calculate the distance between the centers of two balls:

    distance = math.sqrt((x2 - x1)**2 + (y2 - y1)**2)

We will be provided (x1,x2) and (y1,y2) as the position of the centers of two balls.

Than we compare it to the sum of radii of two balls (r1+r2).

In [ ]:
import math
def check_collision(ball1, ball2):
    # Extract out the coordinates and radii
    x1, y1, r1 = ball1
    x2, y2, r2 = ball2

    # to Calculate the distance between the centers
    distance = math.sqrt((x2 - x1)**2 + (y2 - y1)**2)

    # Check for the collision condition
    if distance <= (r1 + r2):
        return True  # The balls are colliding
    else:
        return False  # The balls are not colliding

# Example Test cases
ball1 = [0, 0, 1]
ball2 = [2, 0, 1]
ball3 = [3, 3, 1]
ball4 = [1, 1, 3]
ball5 = [5, 0, 1]

print(check_collision(ball1, ball2))  # will True
print(check_collision(ball1, ball3))  # will False
print(check_collision(ball4, ball5))  # will True


True
False
False


**Answer 1.7**

This problem requires fixing a Python function that calculates the score of a given word according to the given letter values.
The word scoring system involves the sum of the score for each letter while
also adjusting it with length factor in order to have a fairer scoring.
It also comes down to dealing with different scenarios for instance, when the input contains upper and lower cases which should be accommodated in the function to be created in order to make the function more or less case sensitive.

**Explanation:**

•	Using some standardized letter value compile a Python function that is capable of determining the word score.
•	Add up the sum of the letter scores averaged towards an individual word-tabulated total by incorporating an equitable letter-quota score along with the total word count.
•	Make sure that the function is able to handle both lower case inputs and capital letters This will enable the function to be strong enough to meet every need despite the type of input to it being capitalized or written in small letters.


In [ ]:
# Letter values as provided
LETTER_VALUES = {
    'a': 1, 'b': 1, 'c': 5, 'd': 6, 'e': 1, 'f': 4, 'g': 2, 'h': 4, 'i': 1, 'j': 8, 'k': 6,
    'l': 1, 'm': 4, 'n': 1, 'o': 1, 'p': 3, 'q': 11, 'r': 1, 's': 1, 't': 1, 'u': 1, 'v': 5,
    'w': 4, 'x': 8, 'y': 6, 'z': 10, ' ': 0
}

def get_word_score(word):
    """
    Returns the score for a word. Assumes the word is a valid word.

    The score for a word is the product of two components:
        - The first component is the sum of the points for letters in the word.
        - The second component is the larger of:
            1, or
            7*wordlen - 3*(20-wordlen),
          where wordlen is the length of the word.

    Letters are scored as in LETTER_VALUES.

    word: string
    returns: int >= 0
    """
    word = word.lower()  # Convert the word to lowercase

    # Calculate the sum of points for letters in the word
    first_component = sum(LETTER_VALUES[letter] for letter in word)

    # Calculate the length of the word
    wordlen = len(word)

    # Calculate the second component
    second_component = max(1, 7 * wordlen - 3 * (20 - wordlen))

    # The score for the word is the product of the two components
    return first_component * second_component

# Test cases
print(get_word_score("Australia"))  # Example test case 1
print(get_word_score("Deakin"))     # Example test case 2
print(get_word_score("Harsh"))   # Example test case 3

270
16
11


**Code Explanation:**

Let's discuss how the scores are calculated for each of the test cases using the logic provided in the code:

1. "Australia":

Explanation:
First Component:
The word "Australia" consists of the letters: A, u, s, t, r, a, l, i, a.Utilizing genuine strength, tactics, research, and analysis.

Using the LETTER_VALUES, the points for each letter are:
a:1, u:1, s:1, t:1, r:1, a:1, l:1, i:1, a:1

sum of these points = 1 + 1 + 1 + 1 + 1 + 1 + 1 + 1 + 1 = 9

Second Component:

The word “Australia” s length is 9 (wordlen).
Calculate 7 x wordlen - 3  x (20 - wordlen):
Seven times nine is equal to sixty three.
From what was subtracted 9 from 20, it is equal to 11, and multiplying by 3 produces 33.
Second component = 70 - 40 = 30
As 30 is greater than 1 then the second component will be 30.

Final Score:
First Component x Second Component = 9 * 30 = 270

2. "Deakin":

Explanation:
First Component:
The word "Deakin" consists of the letters: Figure 1 shows that the most frequently used letters are D, e, a, k, i and n.

Using the LETTER_VALUES, the points for each letter are:Using the LETTER_VALUES, the points for each letter are:
d : 6, e : 1, a : 1, k : 6, i : 1, n : 1

Total of these points = 6 + 1 + 1 + 6 + 1 + 1 = 16

Second Component:

The word ‘Deakin’ has 6 characters (wordlen).

Calculate 7  wordlen - 3  (20 - wordlen):

This makes it 7 multiplied by 6 which equals to 42
the first one is (20-6=14) while second is 3 * 14 = 42
the second component is : 42 - 42 = 0
This is because the second component of the EWMA formula is equal to 1 as 1 > 0.
 Final SCore = 16 * 1 = 16

**Answer 1.8** - The given figure is for Left Triangle Pascal's pattern, A triangle is said to be left-angled if it has an angle equal to 90 degrees on its left side. To execute this using Python programming, we will be using two for loops nested within an outer for loop: An outer loop to handle the number of rows.

In [ ]:
rows = 5 # number of rows to be printed in star pattern
i = 1
while i <= rows: # Create a loop
    j = i
    while j < rows:
        # display single blank space
        print(' ', end=' ')
        j += 1
    k = 1
    while k <= i:
        print('*', end=' ')
        k += 1
    print()
    i += 1

i = rows # Create a nested loop
while i >= 1:
    j = i
    while j <= rows:
        print(' ', end=' ')
        j += 1
    k = 1
    while k < i:
        print('*', end=' ')
        k += 1
    print('')
    i -= 1

        * 
      * * 
    * * * 
  * * * * 
* * * * * 
  * * * * 
    * * * 
      * * 
        * 
          


**Algo for Left Triangle Pascal's Pattern**

1. Take a number of rows to be printed, lets assume it to be n.
2. Make outer iteration i from 1 to n times to print the rows.
3. Make inner iteration for j from 1 to (N).
4. Print single blank space ” “.
5. Close inner loop (j loop) //its needed for left spacing.
6. Make inner iteration for j from 0 to i.